booster/library/Booster/Pattern/ApplyEquations.hs

```diff
@@ -431,13 +440,25 @@ evaluatePattern' ::
     Pattern ->
     EquationT io Pattern
 evaluatePattern' pat@Pattern{term, ceilConditions} = withPatternContext pat $ do
-    newTerm <- withTermContext term $ evaluateTerm' BottomUp term
+    newTerm <- withTermContext term $ evaluateTerm' BottomUp term `catch_` keepTopLevelResults
     -- after evaluating the term, evaluate all (existing and
     -- newly-acquired) constraints, once
     traverse_ simplifyAssumedPredicate . predicates =<< getState
     -- this may yield additional new constraints, left unevaluated
     evaluatedConstraints <- predicates <$> getState
     pure Pattern{constraints = evaluatedConstraints, term = newTerm, ceilConditions}
+  where
+    -- when TooManyIterations exception occurred while evaluating the top-level term,
+    -- i.e. not in a recursive evaluation of a side-condition,
+    -- it is safe to keep the partial result and ignore the exception.
+    -- Otherwise we would be throwing away useful work.
+    keepTopLevelResults :: LoggerMIO io => EquationFailure -> EquationT io Term
+    keepTopLevelResults = \case
+        err@(TooManyIterations recursionDepth _ _ partialResult) ->
+            case recursionDepth of
+                0 -> pure partialResult
+                _ -> throw err
+        err -> throw err
```

Comment:
After checking the code where we recurse into the evaluation, https://github.com/runtimeverification/haskell-backend/blob/034078d5f2d5f68dada7fcf98a5211eae51bae8f/booster/library/Booster/Pattern/ApplyEquations.hs#L931-L939
I think it is probably fine without this additional `catch_`. Of course it is safer the way you propose it.

Reply:
We still need the `catch_` to get the simplified term, but I do agree that the recursion depth check is redundant. I've removed it.